### PR TITLE
Fixed #28862 Removing a field from index_together/unique_together and from the model generates a migration that crashes

### PR DESCRIPTION
**Pull request by jeffyancey**
update optimizer to correctly order RemoveField after any necessary removal restraints.

### CODE REVIEW

**Comment by charettes** [comment]
Hey @jeffyancey! While this seems to be working, you tests prove it, I don't think passing down models is the appropriate way to tackle the problem.

I believe the appropriate way would be to either adjust the `references_*` methods of the operations or tweak the auto-detector to generate them in the right order from in the first place.

**Comment by jeffyancey** [comment]
Hello charettes, it seems the operations are generated in the correct order, but the optimizer was incorrectly reordering them. I originally tried to correct it in the reduce method, but that also seemed to require passing the models in. I'll see if I can find what you mean by the references, as I would also rather avoid passing the state all the way down to the optimizer if I can find a better way.

File: tests/migrations/test_autodetector.py [hunk 1]
```diff
@@ -1495,6 +1499,20 @@ def test_remove_field_and_foo_together(self):
...
+        changes = self.get_changes(
+            [self.author_empty, self.book_foo_together], [self.author_empty, self.book_foo_together_5]
+        )
+        # Right number/type of migrations?
+        self.assertNumberMigrations(changes, "otherapp", 1)
+        self.assertOperationTypes(changes, "otherapp", 0, ["AlterUniqueTogether", "AlterIndexTogether", "RemoveField"])
+        self.assertOperationAttributes(changes, "otherapp", 0, 0, name="book", unique_together=set())
+        self.assertOperationAttributes(changes, "otherapp", 0, 1, name="book", index_together=set())
+        self.assertOperationAttributes(changes, "otherapp", 0, 2, model_name="book", name="title")
+
```
Comment (charettes) [inline tests/migrations/test_autodetector.py]: By the way, I just checked out your branch locally and you might want to adjust this test as it passes against master without your `django.db.migrations.optimizer` changes.

Comment (jeffyancey) [reply tests/migrations/test_autodetector.py]: Ah, thanks for pointing that out. I was originally piggybacking on a separate test I'll go back and correct it tomorrow.

**Comment by jeffyancey** [comment]
hello @charettes I corrected the optimizer patch, i had a missing check. I've been experimenting with using "references_field", but cannot see how it will help with this particular problem: I need to know what the unique_together, index_together, order_with_respect to sets were for the state prior to any of the existing operations taking effect. If I am not seeing something, any guidance is appreciated, as I'd rather not have to pass the model state through to the optimizer either.

**Comment by edmorley** [comment]
There are now conflicts here after PR #7999 landed btw :-)

**Comment by jeffyancey** [comment]
Thanks for pointing it out. I'll take care of resolving soon.

**Comment by charettes** [comment]
Hey @jeffyancey, sorry from stealing this from you. I know you've invested a lot of effort in getting this fixed and I should have reached out to discuss why I think this a more appropriate solution.

I'd appreciate to get your review on #10178 if you're still available.

**Pull request closed without merging**